stablehlo/transforms/VhloToVersion.cpp

```diff
@@ -78,59 +87,90 @@ FailureOr<Version> validateTargetVersion(llvm::StringRef versionRef,
+                            << " to version " << targetVersion << '\n');
+    return failure();
   }
 
-  auto attrInterface = dyn_cast<VersionedAttrInterface>(attr.getValue());
-  if (attrInterface && isLegalVersion(attrInterface, targetVersion)) {
-    return success();
+  // Recursively check attrs if VHLO attr is a container
+  if (auto arrAttr = attr.dyn_cast<ArrayV1Attr>()) {
+    return success(llvm::all_of(arrAttr.getValue(), [&](Attribute ele) {
+      return succeeded(isLegalAttribute(ele, targetVersion));
+    }));
+  }
+  if (auto elementsAttr = attr.dyn_cast<DenseIntOrFPElementsV1Attr>()) {
+    return isLegalType(elementsAttr.getType(), targetVersion);
+  }
+  if (auto flatSymAttr = attr.dyn_cast<FlatSymbolRefV1Attr>()) {
+    return isLegalAttribute(flatSymAttr.getRootReference(), targetVersion);
+  }
+  if (auto floatAttr = attr.dyn_cast<FloatV1Attr>()) {
+    return isLegalType(floatAttr.getType(), targetVersion);
   }
 
-  LLVM_DEBUG(llvm::dbgs() << "failed to legalize attribute " << attr.getName()
-                          << " to version " << targetVersion);
-  return failure();
+  // Is VHLO and valid version, success.
+  return success();
 }
 
 LogicalResult isLegalType(Type type, const Version& targetVersion) {
-  // TODO: Remove once builtin types are forked.
-  if (isFromDialect(type, "builtin") || isFromDialect(type, "shape")) {
-    return success();
+  // All valid VHLO types must have versioned type interface.
+  auto typeInterface = dyn_cast<VersionedTypeInterface>(type);
+  if (!typeInterface || !isLegalVersion(typeInterface, targetVersion)) {
+    LLVM_DEBUG(llvm::dbgs() << "failed to legalize type " << type
+                            << " to version " << targetVersion << '\n');
+    return failure();
   }
 
-  auto typeInterface = dyn_cast<VersionedTypeInterface>(type);
-  if (typeInterface && isLegalVersion(typeInterface, targetVersion)) {
-    return success();
+  // Recursively check types if VHLO type is a container.
+  if (auto complex = type.dyn_cast<ComplexV1Type>()) {
+    return isLegalType(complex.getElementType(), targetVersion);
+  }
+  if (auto ranked = type.dyn_cast<RankedTensorV1Type>()) {
+    auto encoding = ranked.getEncoding();
+    if (encoding && failed(isLegalAttribute(encoding, targetVersion)))
+      return failure();
+    return isLegalType(ranked.getElementType(), targetVersion);
+  }
+  if (auto quant = type.dyn_cast<UniformQuantizedV1Type>()) {
+    return success(
+        succeeded(isLegalType(quant.getStorageType(), targetVersion)) &&
+        succeeded(isLegalType(quant.getExpressedType(), targetVersion)));
+  }
+  if (auto unranked = type.dyn_cast<UnrankedTensorV1Type>()) {
+    return isLegalType(unranked.getElementType(), targetVersion);
+  }
+  if (auto tuple = type.dyn_cast<TupleV1Type>()) {
+    return success(llvm::all_of(tuple.getTypes(), [&](Type ele) {
+      return succeeded(isLegalType(ele, targetVersion));
+    }));
```

Comment:
If this is supposed to be ordered alphabetically, `TupleV1Type` is out of place.